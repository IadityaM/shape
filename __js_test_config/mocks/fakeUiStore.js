import { observable } from "mobx"

const fakeUiStore = {
  gridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  defaultGridSettings: {
    cols: 4,
    gutter: 20,
    gridW: 312,
    gridH: 250,
  },
  blankContentToolState: {
    order: null,
    width: null,
    height: null,
    replacingId: null,
    placeholderCard: null,
  },
  dialogConfig: {
    open: null,
    prompt: null,
    onConfirm: null,
    onCancel: null,
    iconName: null,
    confirmText: 'OK',
    cancelText: 'Cancel',
    onClose: jest.fn(),
  },
  activityLogOpen: false,
  activityLogPosition: {
    x: 0,
    y: 0,
    h: 1,
    w: 1,
  },
  activityLogMoving: false,
  cardMenuOpen: {
    id: false,
    x: 0,
    y: 0,
    offsetX: 0,
    offsetY: 0,
  },
  selectedArea: { minX: null, maxX: null, minY: null, maxY: null },
  openBlankContentTool: jest.fn(),
  closeBlankContentTool: jest.fn(),
  setBctPlaceholderCard: jest.fn(),
  closeCardMenu: jest.fn(),
  openCardMenu: jest.fn(),
  closeMoveMenu: jest.fn(),
  resetSelectionAndBCT: jest.fn(),
  resetCardPositions: jest.fn(),
  rolesMenuOpen: false,
  isLoading: false,
  dismissedMoveHelper: false,
  selectedAreaEnabled: false,
  setSelectedArea: jest.fn(),
  selectedCardIds: [],
  selectCardId: jest.fn(),
  deselectCards: jest.fn(),
  setViewingRecord: jest.fn(),
  viewingCollection: null,
  isViewingHomepage: false,
  viewingItem: null,
  movingFromCollectionId: null,
  movingCardIds: [],
  openMoveMenu: jest.fn(),
  update: jest.fn(),
  alert: jest.fn(),
  alertOk: jest.fn(),
  defaultAlertError: jest.fn(),
  confirm: jest.fn(),
  closeDialog: jest.fn(),
  cardAction: 'move',
  blurContent: false,
  organizationMenuPage: 'organizationMenuPage',
  organizationMenuGroupId: null,
  organizationMenuOpen: false,
  expandedThreadKey: null,
  zoomLevel: 1,
  expandThread: jest.fn(),
  openGroup: jest.fn(),
  openOptionalMenus: jest.fn(),
  trackEvent: jest.fn(),
  trackedRecords: {},
  editingName: observable([]),
  activityLogPage: 'comments',
  pageMenuOpen: false,
  searchText: '',
  collectionCardSortOrder: '',
  addNewCard: jest.fn(),
  removeNewCard: jest.fn(),
  isNewCard: jest.fn(),
  autocompleteMenuClosed: jest.fn(),
  captureKeyboardGridClick: jest.fn(),
  popupAlert: jest.fn(),
  stopDragging: jest.fn(),
  multiMoveCardIds: [],
  setSnoozeChecked: jest.fn(),
  scrollToTop: jest.fn(),
  scrollToBottom: jest.fn(),
  scrollToBottomOfComments: jest.fn(),
  scrollToBottomOfModal: jest.fn(),
  popupSnackbar: jest.fn(),
  showPermissionsAlert: jest.fn(),
  gridHeightFor: jest.fn().mockReturnValue(250),
  performActionAfterRoute: jest.fn(),
  linkedBreadcrumbTrailForRecord: jest
    .fn()
    .mockImplementation(x => x.breadcrumb),
  addEmptySpaceClickHandler: jest.fn(),
  removeEmptySpaceClickHandler: jest.fn(),
  setEditingCardCover: jest.fn(),
  adminAudienceMenuOpen: false,
  actionMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  textMenuOpenForCard: jest.fn(true).mockImplementation(x => false),
  replyingToCommentId: false,
  setReplyingToComment: jest.fn(),
  setCommentingOnRecord: jest.fn(),
  setActivityLogPage: jest.fn(),
  setBodyBackgroundImage: jest.fn(),
  setBodyFontColor: jest.fn(),
  selectedTextRangeForCard: jest.fn(),
  reselectCardIds: jest.fn(),
  reselectOnlyEditableRecords: jest.fn(),
  reselectOnlyMovableCards: jest.fn(),
  setMovingCards: jest.fn(),
  setVisibleCols: jest.fn(),
  setVisibleRows: jest.fn(),
  setPlaceholderSpot: jest.fn(),
  setDroppingFilesCount: jest.fn(),
  drag: jest.fn(),
  startDragging: jest.fn(),
  adjustZoomLevel: jest.fn(),
  zoomIn: jest.fn(),
  zoomOut: jest.fn(),
  determineZoomLevels: jest.fn(),
  maxCols: jest.fn(),
  maxGridWidth: jest.fn().mockReturnValue(1384),
  pageMargins: jest.fn().mockReturnValue({ left: 100, top: 50 }),
  relativeZoomLevel: 2,
  zoomLevels: [],
  isSelected: jest.fn(),
  toggleSelectedCardId: jest.fn(),
  openContextMenu: jest.fn(),
  clearTextEditingItem: jest.fn(),
  positionForCoordinates: jest.fn().mockReturnValue({ x:0, y:0, xPos: 0, yPos: 0, width: 1, height: 1 }),
  dragGridSpot: new Map(),
  droppingFilesCount: 0,
}

export default fakeUiStore
