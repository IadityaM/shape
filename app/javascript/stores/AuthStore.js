import { types } from 'mobx-state-tree'

// define User model
const User = types
  .model('User', {
    id: types.identifier(types.number),
    email: types.string,
    firstName: types.string,
    lastName: types.string,
  })
  .views(self => ({
    fullName() {
      return `${this.firstName} ${this.lastName}`
    }
  }))
  .actions(self => ({
    changeFirstName(firstName) {
      self.firstName = firstName
    }
  }))

// define model for storing User model
const AuthStore = types
  .model('AuthStore', {
    currentUser: types.maybe(User)
  })
  .actions(self => ({
    loadUser(userData) {
      self.currentUser = userData
    }
  }))

export default AuthStore
