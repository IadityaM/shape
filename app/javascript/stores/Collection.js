import { Record } from 'mobx-jsonapi-store'

class Collection extends Record {}
Collection.type = 'collections'

export default Collection
