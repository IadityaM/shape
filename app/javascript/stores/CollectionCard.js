import BaseRecord from './BaseRecord'

class CollectionCard extends BaseRecord {
  record() {
    if (this.item) {
      return this.item
    } else if (this.collection) {
      return this.collection
    }
    return null
  }
}
CollectionCard.type = 'collection_cards'

export default CollectionCard
