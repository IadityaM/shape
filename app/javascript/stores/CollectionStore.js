import { types } from 'mobx-state-tree'

const fakeCollections = [
  { id: 1, name: 'x' },
  { id: 2, name: 'y' },
  { id: 3, name: 'z' },
  { id: 4, name: 'zzzz' },
]

const fakeCollection = {
  id: 1,
  name: 'My Collection',
  cards: [
    {
      id: 1,
      type: 'atom',
      name: 'Sub-item',
      order: 0,
      width: 1,
      height: 1,
      record: {
        id: 999,
        name: 'Heyo'
      }
    },
  ]
}

// define Collection models
export const CollectionCard = types
  .model('CollectionCard', {
    id: types.identifier(types.number),
    record: types.model({
      id: types.number,
      name: types.string
    }),
    name: types.string,
    type: types.string,
    order: types.number,
    width: types.number,
    height: types.number,
  })

export const Collection = types
  .model('Collection', {
    id: types.identifier(types.number),
    name: types.string,
    cards: types.optional(types.array(CollectionCard), []),
    // type: types.maybe(types.number),
  })

// define model for storing User model
export const CollectionStore = types
  .model('CollectionStore', {
    loading: types.optional(types.boolean, false),
    collections: types.optional(types.array(Collection), []),
    collection: types.maybe(Collection),
  })
  .actions(self => ({
    setCollections(data) {
      self.collections = data
    },
    setLoading(value) {
      self.loading = value
    },
    loadCollections(data) {
      self.loading = true
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          if (data) {
            self.setCollections(data)
          } else {
            self.setCollections(fakeCollections)
          }
          self.setLoading(false)
        }, 1000)
        resolve('hi')
      })
    },
    fetchCollection(id) {
      self.collection = fakeCollection
    }
  }))

export default CollectionStore
