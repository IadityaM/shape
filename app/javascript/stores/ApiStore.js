import { action, observable, computed } from 'mobx'
import { Store } from 'mobx-jsonapi-store'
import _ from 'lodash'

import Collection from './jsonApi/Collection'
import CollectionCard from './jsonApi/CollectionCard'
import Role from './jsonApi/Role'
import FilestackFile from './jsonApi/FilestackFile'
import Group from './jsonApi/Group'
import Item from './jsonApi/Item'
import Organization from './jsonApi/Organization'
import User from './jsonApi/User'

class ApiStore extends Store {
  @observable currentUserId = null

  @action setCurrentUserId(id) {
    this.currentUserId = id
  }

  @computed get currentUser() {
    return this.find('users', this.currentUserId)
  }

  get currentUserOrganization() {
    if (!this.currentUser.current_organization) return null
    return this.currentUser.current_organization
  }

  get currentUserOrganizationId() {
    if (!this.currentUser.current_organization) return null
    return this.currentUser.current_organization.id
  }

  get currentUserOrganization() {
    return this.currentUser.current_organization
  }

  __updateRelationships(obj) {
    const record = this.find(obj.type, obj.id)
    const refs = obj.relationships ? Object.keys(obj.relationships) : []
    refs.forEach((ref) => {
      const items = obj.relationships[ref].data
      if (items instanceof Array && items.length < 1) {
        /* NOTE: special case, if relationship data comes back with an empty array
         * we have to manually empty the array, while also assigning the proper type
         */
        const possibleTypes = _.map(ApiStore.types, model => model.type)
        if (possibleTypes.indexOf(ref) > -1) {
          record.assignRef(ref, observable([]), ref)
        }
      }
    })
    super.__updateRelationships(obj)
  }
}
ApiStore.types = [
  Collection,
  CollectionCard,
  FilestackFile,
  Group,
  Item,
  Role,
  Organization,
  User,
]

export default ApiStore
