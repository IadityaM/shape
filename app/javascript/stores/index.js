import { config as jsonApiConfig, Store } from 'mobx-jsonapi-store'
import { RouterStore } from 'mobx-react-router'

import locale from './Locale'
import User from './User'
import Collection from './Collection'
import Item from './Item'
import CollectionCard from './CollectionCard'

jsonApiConfig.baseUrl = '/api/v1/'
// modify fetch to include 'same-origin' credentials
jsonApiConfig.fetchReference = (url, opts) => {
  opts.credentials = 'same-origin'
  return fetch(url, opts)
}

export const routingStore = new RouterStore()

class ApiStore extends Store {}
ApiStore.types = [User, Collection, Item, CollectionCard]

export const apiStore = new ApiStore()

export default {
  routingStore,
  apiStore,
  // needs to be named "locale"
  locale
}
