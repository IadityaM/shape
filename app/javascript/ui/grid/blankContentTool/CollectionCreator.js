import _ from 'lodash'
import PropTypes from 'prop-types'

import Button from '~/ui/global/Button'
import { BctTextField } from '~/ui/global/styled/forms'
import PaddedCardCover from '~/ui/grid/covers/PaddedCardCover'
import v, { KEYS } from '~/utils/variables'
import { routingStore, uiStore } from '~/stores'
import googleTagManager from '~/vendor/googleTagManager'

class CollectionCreator extends React.Component {
  state = {
    inputText: '',
  }

  onInputChange = e => {
    if (e.target.value.length > v.maxTitleLength) return
    this.setState({
      inputText: e.target.value,
    })
  }

  handleKeyDown = e => {
    if (e.keyCode === KEYS.ESC) {
      this.props.closeBlankContentTool()
    }
  }

  afterCreate = card => {
    googleTagManager.push({
      event: 'formSubmission',
      formType: `Create ${this.dbType || 'Collection'}`,
    })

    // if creating a submissionBox we route you to finish setting up the collection
    if (this.props.type === 'submissionBox')
      routingStore.routeTo('collections', card.record.id)
  }

  createCollection = e => {
    e.preventDefault()
    if (!this.state.inputText) return
    const { createCard, type } = this.props
    const cardParams = {
      // `collection` is the collection being created within the card
      collection_attributes: {
        name: this.state.inputText,
        master_template: this.shouldCreateAsSubTemplate,
        type: this.dbType,
        num_columns: this.numColumns,
      },
    }
    if (type === 'search')
      cardParams.collection_attributes.search_term = this.state.inputText

    createCard(cardParams, {
      afterCreate: this.afterCreate,
    })
  }

  get shouldCreateAsSubTemplate() {
    const { type } = this.props
    const { viewingCollection } = uiStore
    return type === 'template' || viewingCollection.isTemplate
  }

  get dbType() {
    const { type, parentIsFourWide } = this.props

    switch (type) {
      case 'submissionBox':
        return 'Collection::SubmissionBox'
        break
      case 'testCollection':
        return 'Collection::TestCollection'
        break
      case 'foamcoreBoard':
      case 'template':
        return 'Collection::Board'
        break
      case 'search':
        return 'Collection::SearchCollection'
        break
      case 'collection':
        if (parentIsFourWide) return 'Collection::Board'
        return null
    }
    return null
  }

  get numColumns() {
    const { type, parentIsFourWide } = this.props
    if (type === 'foamcoreBoard') {
      return 16
    } else if (
      type === 'template' ||
      (parentIsFourWide && type === 'collection')
    ) {
      return 4
    }
    return null
  }

  get typeName() {
    const { type } = this.props
    // e.g. 'submissionBox' -> 'Submission Box'
    return _.startCase(type)
  }

  render() {
    const { type } = this.props
    const placeholder =
      type === 'search'
        ? 'add search term for collection'
        : `${this.typeName} name`

    return (
      <PaddedCardCover>
        <form className="form" onSubmit={this.createCollection}>
          <BctTextField
            autoFocus
            data-cy="CollectionCreatorTextField"
            placeholder={placeholder}
            value={this.state.inputText}
            onChange={this.onInputChange}
            onKeyDown={this.handleKeyDown}
          />
          <Button
            data-cy="CollectionCreatorFormButton"
            disabled={this.props.loading}
            minWidth={125}
          >
            Create
          </Button>
        </form>
      </PaddedCardCover>
    )
  }
}

CollectionCreator.propTypes = {
  loading: PropTypes.bool.isRequired,
  type: PropTypes.oneOf([
    'collection',
    'template',
    'testCollection',
    'submissionBox',
    'foamcoreBoard',
    'search',
  ]),
  createCard: PropTypes.func.isRequired,
  closeBlankContentTool: PropTypes.func.isRequired,
  parentIsFourWide: PropTypes.bool,
}
CollectionCreator.defaultProps = {
  type: 'collection',
  parentIsFourWide: false,
}

export default CollectionCreator
