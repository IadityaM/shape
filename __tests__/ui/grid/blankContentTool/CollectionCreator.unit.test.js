import CollectionCreator from '~/ui/grid/blankContentTool/CollectionCreator'
import googleTagManager from '~/vendor/googleTagManager'
import { routingStore } from '~/stores'
import fakeUiStore from '#/mocks/fakeUiStore'

jest.mock('../../../../app/javascript/vendor/googleTagManager')
jest.mock('../../../../app/javascript/stores')
const uiStore = fakeUiStore

uiStore.viewingCollection = {
  id: 3,
  isTemplate: false,
}

const e = { preventDefault: jest.fn() }
let wrapper, props, component
const rerender = () => {
  wrapper = shallow(<CollectionCreator {...props} />)
  component = wrapper.instance()
}
describe('CollectionCreator', () => {
  beforeEach(() => {
    props = {
      loading: false,
      type: 'collection',
      createCard: jest.fn(),
      closeBlankContentTool: jest.fn(),
      uiStore: fakeUiStore,
    }
    props.createCard.mockClear()
    rerender()
  })

  it('renders a BctTextField', () => {
    expect(wrapper.find('BctTextField').props().placeholder).toEqual(
      'Collection name'
    )
    expect(wrapper.find('BctTextField').props().autoFocus).toBeTruthy()
  })

  describe('createCollection', () => {
    it('does not call createCard if state.inputText is blank', () => {
      component.state = {
        inputText: '',
      }
      component.createCollection(e)
      expect(props.createCard).not.toHaveBeenCalled()
    })

    it('calls createCard with input name', () => {
      component.state = {
        inputText: 'New Projects',
      }
      component.createCollection(e)
      expect(props.createCard).toHaveBeenCalledWith(
        {
          collection_attributes: {
            name: component.state.inputText,
            master_template: false,
            type: null,
            // collections are 4 columns by default
            num_columns: 4,
          },
        },
        {
          afterCreate: component.afterCreate,
        }
      )
    })

    describe('when collection is a SubmissionBox', () => {
      beforeEach(() => {
        props.type = 'submissionBox'
        props.createCard.mockClear()
        rerender()
      })

      it('creates a card with submission box type and input text name', () => {
        component.state = {
          inputText: 'Challenge #1',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::SubmissionBox',
              num_columns: 4,
            },
          },
          {
            afterCreate: component.afterCreate,
          }
        )
      })
    })

    describe('when collection is a FoamcoreBoard', () => {
      beforeEach(() => {
        props.type = 'foamcoreBoard'
        props.createCard.mockClear()
        rerender()
      })

      it('creates a Collection::Board with 16 columns', () => {
        component.state = {
          inputText: 'Ideas Board',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::Board',
              num_columns: 16,
            },
          },
          {
            afterCreate: component.afterCreate,
          }
        )
      })
    })

    describe('when collection is a SearchColleciton', () => {
      beforeEach(() => {
        props.type = 'search'
        props.createCard.mockClear()
        rerender()
      })

      it('adds the search term to create card', () => {
        component.state = {
          inputText: 'plants',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::SearchCollection',
              search_term: 'plants',
              num_columns: 4,
            },
          },
          {
            afterCreate: component.afterCreate,
          }
        )
      })
    })

    describe('with TestCollection', () => {
      beforeEach(() => {
        props.type = 'testCollection'
        props.createCard.mockClear()
        rerender()
      })

      it('creates a card with test collection type and input text name', () => {
        component.state = {
          inputText: 'My New Test',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::TestCollection',
              num_columns: 4,
            },
          },
          {
            afterCreate: component.afterCreate,
          }
        )
      })
    })

    describe('when a collection is Template', () => {
      beforeEach(() => {
        props.type = 'template'
        props.createCard.mockClear()
        rerender()
      })

      it('creates a 4WFC template collection', () => {
        component.state = {
          inputText: 'My New Template',
        }
        component.createCollection(e)
        expect(component.shouldCreateAsSubTemplate).toBeTruthy()
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: true,
              type: 'Collection::Board',
              num_columns: 4,
            },
          },
          {
            afterCreate: component.afterCreate,
          }
        )
      })
    })

    describe('when a viewing collection is a master template', () => {
      beforeEach(() => {
        props.uiStore.viewingCollection.isTemplate = true
        props.createCard.mockClear()
        rerender()
      })

      it('creates a template collection', () => {
        component.createCollection(e)
        expect(component.shouldCreateAsSubTemplate).toBeTruthy()
      })
    })
  })

  describe('afterCreate', () => {
    describe('when collection is any collection', () => {
      it('pushes an event to google tag manager', () => {
        wrapper.setProps({ type: 'collection' })
        component.afterCreate({})
        expect(googleTagManager.push).toHaveBeenCalledWith({
          event: 'formSubmission',
          formType: 'Create Collection',
        })
      })
    })
    describe('when collection is a submission box', () => {
      it('pushes an event to google tag manager and redirects', () => {
        wrapper.setProps({ type: 'submissionBox' })
        component.afterCreate({
          record: {
            id: 1,
          },
        })

        expect(googleTagManager.push).toHaveBeenCalledWith({
          event: 'formSubmission',
          formType: 'Create Collection::SubmissionBox',
        })
        expect(routingStore.routeTo).toHaveBeenCalledWith('collections', 1)
      })
    })
  })
})
