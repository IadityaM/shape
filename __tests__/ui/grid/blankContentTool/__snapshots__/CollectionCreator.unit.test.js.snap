// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionCreator renders snapshot 1`] = `
<styled.div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <BctTextField
      autoFocus={true}
      data-cy="CollectionCreatorTextField"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Collection name"
      value=""
    />
    <FormButton
      color="#120f0e"
      data-cy="CollectionCreatorFormButton"
      disabled={false}
      fontFamily="'Gotham', sans-serif"
      fontSize={1}
      fontWeight={500}
      transparent={false}
      width={125}
    >
      Create
    </FormButton>
  </form>
</styled.div>
`;
